packages/vhd-lib/src/Vhd/VhdAbstract.js

```diff
@@ -204,4 +213,93 @@ export class VhdAbstract {
     const relativePathToTarget = path.relative(aliasDir, path.resolve('/', targetPath))
     await handler.writeFile(aliasPath, relativePathToTarget)
   }
+
+  stream() {
+    const { header, footer, batSize } = this
+    const rawFooter = fuFooter.pack(footer)
+    checksumStruct(rawFooter, fuFooter)
+
+    // compute parent locator place and size
+    // update them in header
+    // update checksum in header
+
+    let offset = FOOTER_SIZE + HEADER_SIZE + batSize
+    for (let i = 0; i < PARENT_LOCATOR_ENTRIES; i++) {
+      if (header.parentLocatorEntry[i].platformDataSpace > 0) {
+        header.parentLocatorEntry[i].platformDataOffset = offset
+        offset += header.parentLocatorEntry[i].platformDataSpace
+      }
+    }
+
+    const rawHeader = fuHeader.pack(header)
+    checksumStruct(rawHeader, fuHeader)
+
+    assert.strictEqual(offset % SECTOR_SIZE, 0)
+
+    const bat = Buffer.allocUnsafe(batSize)
+    let offsetSector = offset / SECTOR_SIZE
+    const blockSizeInSectors = this.fullBlockSize / SECTOR_SIZE
+
+    // compute BAT , blocks starts after parent locator entries
+    for (let i = 0; i < header.maxTableEntries; i++) {
+      if (this.containsBlock(i)) {
+        bat.writeUInt32BE(offsetSector, i * 4)
+        offsetSector += blockSizeInSectors
+      } else {
+        bat.writeUInt32BE(BLOCK_UNUSED, i * 4)
+      }
+    }
+    const fileSize = offsetSector * SECTOR_SIZE + FOOTER_SIZE /* the footer at the end */
+
+    const self = this
+    async function* iterator() {
+      yield rawFooter
+      yield rawHeader
+      yield bat
+
+      // yield parent locator entries
+      for (let i = 0; i < PARENT_LOCATOR_ENTRIES; i++) {
+        if (header.parentLocatorEntry[i].platformDataSpace > 0) {
+          const parentLocator = await self.readParentLocator(i)
+          // @ todo pad to platformDataSpace
+          yield parentLocator.data
+        }
+      }
+
+      // yield all blocks
+      // since contains() can be costly for synthetic vhd, use the computed bat
+      for (let i = 0; i < header.maxTableEntries; i++) {
+        if (bat.readUInt32BE(i * 4) !== BLOCK_UNUSED) {
+          const block = await self.readBlock(i)
+          yield block.buffer
+        }
+      }
+      // yield footer again
+      yield rawFooter
+    }
+
+    const stream = asyncIteratorToStream(iterator())
+    stream.length = fileSize
+    return stream
+  }
+
+  rawContent() {
+    const { header, sectorsPerBlock } = this
+    const self = this
+
+    async function* iterator() {
+      const nBlocks = header.maxTableEntries
+      const EMPTY = Buffer.allocUnsafe(sectorsPerBlock * SECTOR_SIZE)
+      for (let blockId = 0; blockId < nBlocks; ++blockId) {
+        if (await self.containsBlock(blockId)) {
+          yield (await self.readBlock(blockId)).data
```

Comment:
Not related to this PR but I'm writing this here because I'm thinking about it: could we improve the performance by removing the call to `containsBlock` and let `readBlock` tell us if there is no such block? :thinking:

Reply:
contains block only need to check a few bits in a local Buffer. ReadBlock can by a http query to S3

Reply:
`readBlock` could handle this case itself, either by throwing or returning a special value.

Anyway, it's unrelated to this PR, I'm closing this comment :slightly_smiling_face: